File: python/src/rules.py
"""Rule definitions: what to limit, over which window, at which scope."""

from __future__ import annotations

import json
from dataclasses import dataclass
from enum import Enum


class Scope(str, Enum):
    """Where a limit is enforced. A request is keyed differently per scope so the
    same rule isolates one global fleet, one tenant, or one user."""

    GLOBAL = "global"
    TENANT = "tenant"
    USER = "user"


@dataclass(frozen=True)
class LimitRule:
    """One provider-style limit. Any subset of the max_* dimensions may be set;
    a request must stay under every one that is."""

    model: str
    window_seconds: float
    max_tokens: float | None = None
    max_requests: float | None = None
    max_cost: float | None = None
    max_concurrent: int | None = None
    scope: Scope = Scope.GLOBAL
    buckets_per_window: int = 60
    precise: bool = False
    name: str | None = None

    @property
    def bucket_seconds(self) -> float:
        """0 selects the exact per-event log; otherwise window/buckets_per_window."""
        if self.precise:
            return 0.0
        return self.window_seconds / max(1, self.buckets_per_window)

    @property
    def label(self) -> str:
        return self.name or f"{self.model}:{self.scope.value}:{int(self.window_seconds)}s"


def rules_from_json(text: str) -> list[LimitRule]:
    doc = json.loads(text)
    out: list[LimitRule] = []
    for row in doc["rules"]:
        out.append(
            LimitRule(
                model=row["model"],
                window_seconds=float(row["window_seconds"]),
                max_tokens=row.get("max_tokens"),
                max_requests=row.get("max_requests"),
                max_cost=row.get("max_cost"),
                max_concurrent=row.get("max_concurrent"),
                scope=Scope(row.get("scope", "global")),
                buckets_per_window=int(row.get("buckets_per_window", 60)),
                precise=bool(row.get("precise", False)),
                name=row.get("name"),
            )
        )
    return out

File: ts/src/rules.ts
/** Rule definitions: what to limit, over which window, at which scope. */

/**
 * Where a limit is enforced. A request is keyed differently per scope so the same
 * rule isolates one global fleet, one tenant, or one user.
 */
export enum Scope {
  Global = "global",
  Tenant = "tenant",
  User = "user",
}

/** Maps a wire name to a {@link Scope}, throwing for unknown names. */
export function parseScope(s: string): Scope {
  switch (s) {
    case "global":
      return Scope.Global;
    case "tenant":
      return Scope.Tenant;
    case "user":
      return Scope.User;
    default:
      throw new Error(`unknown scope ${s}`);
  }
}

/** The optional dimensions and tuning knobs of a {@link LimitRule}. */
export interface LimitRuleOptions {
  maxTokens?: number;
  maxRequests?: number;
  maxCost?: number;
  maxConcurrent?: number;
  scope?: Scope;
  bucketsPerWindow?: number;
  precise?: boolean;
  name?: string;
}

/**
 * One provider-style limit. Any subset of the `max*` dimensions may be set; a
 * request must stay under every one that is.
 */
export class LimitRule {
  readonly model: string;
  readonly windowSeconds: number;
  readonly maxTokens?: number;
  readonly maxRequests?: number;
  readonly maxCost?: number;
  readonly maxConcurrent?: number;
  readonly scope: Scope;
  readonly bucketsPerWindow: number;
  readonly precise: boolean;
  readonly name?: string;

  constructor(model: string, windowSeconds: number, opts: LimitRuleOptions = {}) {
    this.model = model;
    this.windowSeconds = windowSeconds;
    this.maxTokens = opts.maxTokens;
    this.maxRequests = opts.maxRequests;
    this.maxCost = opts.maxCost;
    this.maxConcurrent = opts.maxConcurrent;
    this.scope = opts.scope ?? Scope.Global;
    this.bucketsPerWindow = opts.bucketsPerWindow ?? 60;
    this.precise = opts.precise ?? false;
    this.name = opts.name;
  }

  /** 0 selects the exact per-event log; otherwise window/bucketsPerWindow. */
  get bucketSeconds(): number {
    if (this.precise) {
      return 0;
    }
    return this.windowSeconds / Math.max(1, this.bucketsPerWindow);
  }

  /** The rule's name, or a derived `model:scope:Ns` label. */
  get label(): string {
    return this.name ?? `${this.model}:${this.scope}:${Math.trunc(this.windowSeconds)}s`;
  }

  /**
   * Whether the rule caps tokens, requests, or cost (as opposed to being a
   * concurrency-only rule).
   */
  hasUsageLimit(): boolean {
    return this.maxTokens !== undefined || this.maxRequests !== undefined || this.maxCost !== undefined;
  }
}

interface RuleRow {
  model: string;
  window_seconds: number;
  max_tokens?: number;
  max_requests?: number;
  max_cost?: number;
  max_concurrent?: number;
  scope?: string;
  buckets_per_window?: number;
  precise?: boolean;
  name?: string;
}

/** Parses a rules document (see limits.sample.json) into {@link LimitRule}s. */
export function rulesFromJson(text: string): LimitRule[] {
  const doc = JSON.parse(text) as { rules: RuleRow[] };
  return doc.rules.map(
    (row) =>
      new LimitRule(row.model, Number(row.window_seconds), {
        maxTokens: row.max_tokens,
        maxRequests: row.max_requests,
        maxCost: row.max_cost,
        maxConcurrent: row.max_concurrent,
        scope: row.scope === undefined ? Scope.Global : parseScope(row.scope),
        bucketsPerWindow: row.buckets_per_window === undefined ? 60 : Number(row.buckets_per_window),
        precise: row.precise === undefined ? false : Boolean(row.precise),
        name: row.name,
      }),
  );
}
